### DISCUSSION TITLE
How to use GraphSAINTRandomWalkSampler for BipartiteData

### DISCUSSION THREAD

**Discussion by freemanhm** [pyg-team/pytorch_geometric#2587] [Q&A]
Hi all,
I followed the torch_geometric tutorial and I am using Bipartite Graph for link prediction.

    class BipartiteData(Data):
        def __init__(self, edge_index, x_s, x_t):
            super(BipartiteData, self).__init__()
            self.edge_index = edge_index
            self.x_s = x_s
            self.x_t = x_t
    
        def __inc__(self, key, value):
            if key == 'edge_index':
                return torch.tensor([[self.x_s.size(0)], [self.x_t.size(0)]])
            else:
                return super().__inc__(key, value)
we add some attributes to the data(BipartiteData). The size of x_s =[130000,10],x_t[100000,12], It is too big to use a single GPU for training, so I want to split this graph with GraphSAINTRandomWalkSampler().

    data= BipartiteData(edge_index=edge_index,x_s = x_s,x_t=x_t)
    data.train_pos_edge_index = train_pos_edge_index
    data.train_neg_edge_index = train_neg_edge_index
    data.test_pos_edge_index = test_pos_edge_index
    data.user_emb_nums = x_t_emb_num.int()
    data.item_emb_nums = x_s_emb_num.int()
    data.item_title_x = item_title_x
    data.item_tags_x = item_tags_x
    data.user_jobroles_x = user_jobroles_x
    data.words_num = words_num

    loader = GraphSAINTRandomWalkSampler(data, batch_size=6000, walk_length=2, num_steps=5, sample_coverage=100,)
    for data in loader:
	    print(data.x_s.size())

but I get the following problems:

      Traceback (most recent call last):
        File "/root/miniconda3/envs/myconda/lib/python3.8/site-packages/IPython/core/interactiveshell.py", line 3418, in run_code
          exec(code_obj, self.user_global_ns, self.user_ns)
        File "<ipython-input-1-d07b1f27fae1>", line 1, in <module>
          for data in loader:
        File "/root/miniconda3/envs/myconda/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 435, in __next__
          data = self._next_data()
        File "/root/miniconda3/envs/myconda/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 475, in _next_data
          data = self._dataset_fetcher.fetch(index)  # may raise StopIteration
        File "/root/miniconda3/envs/myconda/lib/python3.8/site-packages/torch/utils/data/_utils/fetch.py", line 47, in fetch
          return self.collate_fn(data)
        File "/root/miniconda3/envs/myconda/lib/python3.8/site-packages/torch_geometric/data/graph_saint.py", line 102, in __collate__
          data = self.data.__class__()
      TypeError: __init__() missing 3 required positional arguments: 'edge_index', 'x_s', and 'x_t'

**Comment by rusty1s** [comment]
I see. The error comes from requiring the 3 input arguments of `BipartiteData`. Dropping the `__init__` function completely should fix this issue. However, I don't think that `GraphSAINT` is well-defined on bipartite graph-structured data but instead expects to operate in a homogeneous graph. For this, it's best to convert your graph into a homogeneous one first before running `GraphSAINT` samplers (e.g., via `edge_index[:, 1] += 100000`)

**Reply by freemanhm** [reply]
OK, I'll try it according to your suggestion. Thanks a lot for your answer and this awesome repository 👍

**Reply by freemanhm** [reply]
However, there are two types of nodes with different features in my Bipartite Graph. How can I convert the graph into a homogeneous one?

**Reply by rusty1s** [reply]
Yes, this is a bit tricky in that case and we are working on better support for that. One possible workaround is to keep both `x_s` and `x_t` outside `data`, and instead add index attributes to `data` that denote to which type each node belongs, and what's it's global index in either `x_s` or `x_t`, e.g.:
```python
data.n_type = torch.cat([torch.zeros(130000, dtype=torch.long), torch.ones(100000, dtype=torch.long)])
data.n_id = torch.cat([torch.arange(130000), torch.arange(100000)])
```
which can then be used *after* sampling to reconstruct mini-batched `x_s` and `x_t`.

**Reply by freemanhm** [reply]
ok，I will try.

**Reply by freemanhm** [reply]
It works when I revised my codes per your suggestions. Thanks a lot.
Now, I have another question about how I should set the parameters' values in the "GraphSAINTRandomWalkSampler(data, batch_size=6000, walk_length=2, num_steps=5, sample_coverage=100,)". In other words, I am not sure about the relationship between the "batch_size" and "num_steps". In my opinion, I would set the "num_steps=data.num_nodes/batch_size", is that right?
